### DISCUSSION TITLE
MVUX multi threading and race conditions

### DISCUSSION THREAD

**Discussion by igiona** [unoplatform/uno#17345] [Q&A] [answered]
I'm facing some issues (race condition) with the MVUX pattern, which I somehow believe are intrinsically present with the way the pattern is defined.

## The use case:
I have a view which requires 2 values of from a model to display an icon (battery level and charging-status).
These 2 values are populated from 2 distinct events.

The model is as follow:

```cs
public record PowerManagementModel(
    BatteryInformationModel BatteryInformation
);

public record BatteryInformationModel(
    Ratio? BatteryStateOfCharge,
    BatteryChargingState? ChargingState
);

public IState<PowerManagementModel> PowerManagement => State<PowerManagementModel>.Value(this, () => new(new(null, null)));

```
The values is bound  to an Icon in XAML, with a converter:

```xaml
<BitmapIcon UriSource="{Binding PowerManagement.BatteryInformation, Converter={StaticResource BsocToIconSoureConverter}}" VerticalAlignment="Center" Height="22" />
``` 
So far so good... but

### Q1
I need to wrap `BatteryInformationModel` into the `PowerManagementModel`, because otherwise the `UriSource` binding on the `PowerManagement` property doesn't work, is that per design?

### Q2

As said before, the elements of the battery information are generated independently with two handlers:

```cs
    private async Task HandleBsocEventsAsync(BatteryStateOfCharge bsoc, CancellationToken cancellationToken)
    {
        var pm = await PowerManagement ?? throw new InvalidOperationException("Unable to get the current pm model state");
        await PowerManagement.Update(_ => pm with { BatteryInformation = pm.BatteryInformation with { BatteryStateOfCharge = bsoc.Value } }, cancellationToken);
    }

    private async Task HandleChargingEventsAsync(BatteryChargingState chargingState, CancellationToken cancellationToken)
    {
        var pm = await PowerManagement ?? throw new InvalidOperationException("Unable to get the current pm model state");
        await PowerManagement.Update(_ => pm with { BatteryInformation = pm.BatteryInformation with { ChargingState = chargingState } }, cancellationToken);
    }
```

Here comes the problem.
Assume a IState with value 50% and `charging`.
When these two handlers happend concurrently, let's say with the new battery state of charge 60% and the new chargingState `not charging`, bot handler capture the IState (50%, `charging`).
Depending on the execution priority of the `Update` function, I will get either the wrong battery level or charging state :'(
The fact that in order to update a state, I need to capture it before, make it for me hard to avoid this race condition.

What is the idiomatic way to use MVUX in such use case, in order to avoid this race condition?

Thank you in advance!

**Comment by francoistanguay** [comment]
I'm not 100% sure I follow everything, and a repro sample would help but for what it's worth, the IState.UpdateAsync(x => x...) gives you the current value as the input to the lambda exactly to help with concurrency issues.

You'd get a snapshot, return a new value, and if a subsequent call to update is made, you would get that new value as the input.

In general, you should never await the current value and always have access to it in UpdateAsync, you get also get it directly in your public async method as an implicit command parameter.

Also, regarding BatteryModel having to be nested in PowerModel, you can have multiple IState(s) in your Model.

**Reply by dr1rrb** [reply]
#Q1:

Not sure to understand your issue, do you have a repro? You can indeed have multiple states, another hypothesis would be an issue your converter that requests to have an `BatteryInformationModel` while in fact you will get a `***Bindable***BatteryInformationModel`.

#Q2:

As @francoistanguay said, you should not await you state and instead use the "current value" provided in the `Update` callback:

```csharp
    private async Task HandleBsocEventsAsync(BatteryStateOfCharge bsoc, CancellationToken cancellationToken)
    {
        await PowerManagement.UpdateAsync(pm=> (pm ?? new()) with { BatteryInformation = pm.BatteryInformation with { BatteryStateOfCharge = bsoc.Value } }, cancellationToken);
    }

    private async Task HandleChargingEventsAsync(BatteryChargingState chargingState, CancellationToken cancellationToken)
    {
        await PowerManagement.UpdateAsync(pm => (pm ?? new())with { BatteryInformation = pm.BatteryInformation with { ChargingState = chargingState } }, cancellationToken);
    }
```

**Reply by igiona** [reply]
Thank you both for the support, wrt to Q2: I wasn't aware of the fact that the `UpdateAsync` actually provides the current value in the closure.
This indeed helps, many thanks!

@francoistanguay I'm not sure what you mean with 

> [..] you get (can) also get it directly in your public async method as an implicit command parameter.

With regards to Q1, I have a repro here: https://github.com/igiona/uno-repro-binding
I now think that it's a bug, because it works on Windows but not on Android.
The `main` branch doesn't have the nested models and doesn't work, the `working-example` branch adds the nesting and it does work also on Android.
EDIT: with "doesn't work", I mean that the converter is not called when the binding changes.
It's only called once at startup and never again afterwards.

**Reply by francoistanguay** [reply]
When you have a named Feed/State in your Model, there's a convention that lets you receive the latest "snapshot" automatically as a Command Parameter:

class Model
  IState<< Counter >> Count { get; }
  
  void Increment(Counter count) //parameter has same name as a State/Feed property
  {
     //Do something with counter. It's mainly useful if you're calling an API, not super useful when updating the state as the other syntax with the lambda is preferable.
  }

**Reply by igiona** [reply]
Thanks for the clarification 👍

**Reply by dr1rrb** [reply, marked as answer]
Ok so I look in your code, you indeed get a `BindablePowerManagementModel` but you converter is already assuming this. The issue is that once you get it, the instance of `BindablePowerManagementModel` will never change, so you converter will not be invoked again (It's the generated bindable proxy that is holding the state for the view / the binding engine. The instance won't change, it will instead raise "property change" event itself to update the view.)

To fix this you have 3 options:

## Calculated property in the model

Considering the type of converter, you can just remove it and add a computed property in your model:

```csharp
public record PowerManagementModel(
	double? BatteryStateOfCharge,
	BatteryChargingState? ChargingState
)
{
	public static PowerManagementModel Empty => new PowerManagementModel(null, null);
	
	public string Description => $"Level is {BatteryStateOfCharge} and currently {ChargingState}";
}
```

and then change your binding accordingly:

```xml
<TextBox Text="{Binding PowerManagement.Description}" VerticalAlignment="Center" Height="22" />
```

It's one of the advantages of having immutable record / MVUX, you can remove lot of verbose converters like this.

## Projection in the `MainModel`

An issue of the previous solution is that it cannot be easily localized, a solution then you be to use a projection in the `MainModel`, e.g.:

```csharp
public IFeed<string> PowerManagementDescription => PowerManagement.Select(pm => _localizer[Loc_Key, pm.BatteryStateOfCharge, pm.ChargingState]);
```

```xml
<TextBox Text="{Binding PowerManagementDescription}" VerticalAlignment="Center" Height="22" />
```

## Using `.Value` property

Finally for cases where the converter are inevitable (like having a dependency to some UI elements), you can bind on the hidden `Value` property which will expose the current `PowerManagementModel`:

Change your converter to accept a `PowerManagementModel`:
```csharp
var battInfo = value as PowerManagementModel;
if (battInfo != null)
{
    return $"Level is {battInfo.BatteryStateOfCharge} and currently {battInfo.ChargingState}";
}
```

Then change you binding to :

```xml
<TextBox Text="{Binding PowerManagement.Value, Converter={StaticResource BsocToTextConverter}}" VerticalAlignment="Center" Height="22" />
```



Note: I opened https://github.com/unoplatform/uno.extensions/issues/2410 which could be a solution to silently support your case.

**Reply by igiona** [reply]
@dr1rrb thank you for the exhaustive answer 👍 

In reality the converter is a bit more complex than the one in the repro, as it computes the icon-source to be displayed. Of course ti could be moved into the model, but I don't like to mix this UI aspects in the model.

>It's one of the advantages of having immutable record / MVUX, you can remove lot of verbose converters like this.

Can you maybe elaborate more on that?
What's "bad" with the converters and why an immutable record helps? One could expose such property to replace a converter also in MVVM. 
It's performance what you refer to ?

For now, I think I will go for the `.Value` :)

**Reply by francoistanguay** [reply]
One thing is that Converters cannot easily get access to the services you might need like Localizer. 

So sometimes it's easier to put your presentation logic in the MVUX class which is really the equivalent of an Immutable/Reactive ViewModel which to a certain point is where you can put "View/Presentation" logic.

In all cases, .Value is an easy way to go about it. You just have other options that might be valid in certain cases.
